all: implement workspace/reference v2

slimsag opened the pull request:
According to the new specification at https://github.com/sourcegraph/language-server-protocol/pull/8

Please do not merge until that PR is.

keegancsmith commented on langserver/workspace_refs.go:
```diff
@@ -222,19 +209,19 @@ func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Conte
+				// Non-standard fields.
+				"dir":    defPkg.Dir, // TODO: emit standard "file" instead?
+				"pkg":    defPkg.ImportPath,
+				"vendor": IsVendorDir(defPkg.Dir),
```
shouldn't these fields live in meta?

keegancsmith commented on langserver/workspace_refs.go:
```diff
@@ -217,7 +217,6 @@ func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Conte
 				ContainerName: defContainerName,
 				PackageName:   defPkg.ImportPath,
 				Meta: map[string]interface{}{
-					"dir":    defPkg.Dir, // TODO: emit standard "file" instead?
```
surprised you didn't need to update tests with this commit?

Edit: I see you updated in a later commit

slimsag commented:
1. I've updated the `workspace/xreference` implementation to reflect the latest specification.
2. I've also added a `textDocument/xdefinition` implementation according to the spec.
3. I've also squashed this down into just a single commit for `workspace/xreferences` and `textDocument/xdefinition`.

The xreference impl has not changed much since @keegancsmith reviewed it, but the second commit / xdefinition implementation is new so I'll need to get a review on that. @keegancsmith I will ask someone else to review because I want to land this into master today, _but_ please leave comments etc here and I will follow up tomorrow morning.

beyang commented on langserver/definition.go:
```diff
@@ -40,11 +58,32 @@ func (h *LangHandler) handleDefinition(ctx context.Context, conn JSONRPC2Conn, r
+		l.Location.Range.End = l.Location.Range.Start
+
+		// Determine metadata information for the node.
+		//rootFile := pkg.Files[0] // TODO: is this correct?
```
remove commented-out code

slimsag replied:
(hint: it was _not_ correct :) )

beyang commented on langserver/definition.go:
```diff
@@ -40,11 +58,32 @@ func (h *LangHandler) handleDefinition(ctx context.Context, conn JSONRPC2Conn, r
+
+		// Determine metadata information for the node.
+		//rootFile := pkg.Files[0] // TODO: is this correct?
+		def, err := refs.DefInfo(pkg.Pkg, &pkg.Info, pathEnclosingInterval, node.Pos())
```
Since `def` is only used within the following if-else block, a clearer way to write this would be:

```
if def, err := refs.DefInfo(pkg.Pkg, &pkg.Info, pathEnclosingInterval, node.Pos()); err == nil {

} else 

}
```

beyang commented on langserver/workspace_refs.go:
```diff
@@ -189,10 +180,6 @@ func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Conte
 	}()
 	span.SetTag("pkg", pkg)
 
-	pkgInWorkspace := func(path string) bool {
```
If we now emit references to both internal and external definitions, you should change the name and docstring of this function.

slimsag replied:
(done)

beyang commented on pkg/lspext/lspext.go:
```diff
@@ -1,47 +1,80 @@
+	Symbol    SymbolDescriptor `json:"symbol"`
+}
+
+type SymbolDescriptor struct {
```
Please add a docstring for this type and each of the fields.

beyang commented on pkg/lspext/lspext.go:
```diff
@@ -1,47 +1,80 @@
+//
+// See: https://github.com/sourcegraph/language-server-protocol/blob/master/extension-workspace-reference.md
+//
 type ReferenceInformation struct {
```
Please add a docstring for each of these fields.

beyang commented on pkg/lspext/lspext.go:
```diff
@@ -1,47 +1,80 @@
+	File          string                 `json:"file,omitempty"`
+	ContainerName string                 `json:"containerName,omitempty"`
+	Vendor        bool                   `json:"vendor,omitempty"`
+	Meta          map[string]interface{} `json:"meta,omitempty"`
```
If `SymbolDescriptor` itself is "metadata", having a `Meta` field inside it seems redundant. What about something like `Attributes`?

slimsag replied:
That would be a spec change, too. But okay.

slimsag commented:
Also see: https://github.com/sourcegraph/sourcegraph/pull/2673
